Panic on inserting a newline

### Summary

​

### Reproduction Steps

```bash
$ echo 'if(true) &&true' > problem.ts
$ hx problem.ts
8li<ret>
```

### Helix log

Backtrace:
```
thread 'main' panicked at 'called `Option::unwrap()` on a `None` value', helix-core/src/indent.rs:516:60
stack backtrace:
   0: rust_begin_unwind
             at /rustc/fe5b13d681f25ee6474be29d748c65adcd91f69e/library/std/src/panicking.rs:584:5
   1: core::panicking::panic_fmt
             at /rustc/fe5b13d681f25ee6474be29d748c65adcd91f69e/library/core/src/panicking.rs:143:14
   2: core::panicking::panic
             at /rustc/fe5b13d681f25ee6474be29d748c65adcd91f69e/library/core/src/panicking.rs:48:5
   3: helix_core::indent::treesitter_indent_for_pos
   4: helix_core::indent::indent_for_newline
   5: core::ops::function::impls::<impl core::ops::function::FnOnce<A> for &mut F>::call_once
   6: helix_core::transaction::Transaction::change_by_selection
   7: helix_term::commands::insert::insert_newline
   8: helix_term::ui::editor::EditorView::handle_keymap_event::{{closure}}
   9: helix_term::ui::editor::EditorView::handle_keymap_event
  10: helix_term::ui::editor::EditorView::insert_mode
  11: <helix_term::ui::editor::EditorView as helix_term::compositor::Component>::handle_event
  12: helix_term::compositor::Compositor::handle_event
  13: helix_term::application::Application::handle_terminal_events
  14: hx::main_impl::{{closure}}
  15: std::thread::local::LocalKey<T>::with
  16: tokio::park::thread::CachedParkThread::block_on
  17: tokio::runtime::scheduler::multi_thread::MultiThread::block_on
  18: tokio::runtime::Runtime::block_on
  19: hx::main
note: Some details are omitted, run with `RUST_BACKTRACE=full` for a verbose backtrace.
```
Log:
```
2022-11-03T13:02:19.048 helix_lsp::transport [ERROR] err: <- StreamClosed
2022-11-03T13:02:19.048 helix_lsp::transport [ERROR] err: <- StreamClosed
2022-11-03T13:02:24.536 helix_loader [DEBUG] Located configuration folders: []
2022-11-03T13:02:24.542 helix_view::clipboard [INFO] Using xclip to interact with the system and selection (primary) clipboard
2022-11-03T13:02:24.658 mio::poll [TRACE] registering event source with poller: token=Token(1), interests=READABLE | WRITABLE
2022-11-03T13:02:24.658 mio::poll [TRACE] registering event source with poller: token=Token(2), interests=READABLE | WRITABLE
2022-11-03T13:02:24.658 mio::poll [TRACE] registering event source with poller: token=Token(3), interests=READABLE | WRITABLE
2022-11-03T13:02:24.658 helix_view::editor [DEBUG] editor status: Loaded 1 files.
2022-11-03T13:02:24.658 mio::poll [TRACE] registering event source with poller: token=Token(4), interests=READABLE | WRITABLE
2022-11-03T13:02:24.658 mio::poll [TRACE] registering event source with poller: token=Token(5), interests=READABLE | WRITABLE
2022-11-03T13:02:24.658 helix_lsp::transport [INFO] -> {"jsonrpc":"2.0","method":"initialize","params":{"capabilities":{"textDocument":{"codeAction":{"codeActionLiteralSupport":{"codeActionKind":{"valueSet":["","quickfix","refactor","refactor.extract","refactor.inline","refactor.rewrite","source","source.organizeImports"]}}},"completion":{"completionItem":{"resolveSupport":{"properties":["documentation","detail","additionalTextEdits"]},"snippetSupport":false},"completionItemKind":{}},"hover":{"contentFormat":["markdown"]},"publishDiagnostics":{},"rename":{"dynamicRegistration":false,"honorsChangeAnnotations":false,"prepareSupport":false},"signatureHelp":{"signatureInformation":{"activeParameterSupport":true,"documentationFormat":["markdown"],"parameterInformation":{"labelOffsetSupport":true}}}},"window":{"workDoneProgress":true},"workspace":{"applyEdit":true,"configuration":true,"didChangeConfiguration":{"dynamicRegistration":false},"symbol":{"dynamicRegistration":false},"workspaceFolders":true}},"processId":3283485,"rootPath":"/home/itay123","rootUri":"file:///home/itay123","workspaceFolders":[{"name":"itay123","uri":"file:///home/itay123"}]},"id":0}
2022-11-03T13:02:24.658 mio::poll [TRACE] registering event source with poller: token=Token(0), interests=READABLE
2022-11-03T13:02:24.658 mio::poll [TRACE] registering event source with poller: token=Token(1), interests=READABLE
2022-11-03T13:02:24.658 helix_view::document [DEBUG] id 1 modified - last saved: 0, current: 0
2022-11-03T13:02:24.850 helix_lsp::transport [INFO] <- {"jsonrpc":"2.0","method":"window/logMessage","params":{"type":3,"message":"[lspserver] Using Typescript version (bundled) 4.8.4 from path \"/usr/lib/node_modules/typescript/lib/tsserver.js\""}}
2022-11-03T13:02:24.850 helix_term::application [DEBUG] received editor event: LanguageServerMessage((0, Notification(Notification { jsonrpc: Some(V2), method: "window/logMessage", params: Map({"message": String("[lspserver] Using Typescript version (bundled) 4.8.4 from path \"/usr/lib/node_modules/typescript/lib/tsserver.js\""), "type": Number(3)}) })))
2022-11-03T13:02:24.850 helix_term::application [INFO] window/logMessage: LogMessageParams { typ: Info, message: "[lspserver] Using Typescript version (bundled) 4.8.4 from path \"/usr/lib/node_modules/typescript/lib/tsserver.js\"" }
2022-11-03T13:02:24.850 helix_lsp::transport [INFO] <- {"jsonrpc":"2.0","id":0,"result":{"capabilities":{"textDocumentSync":2,"completionProvider":{"triggerCharacters":[".","\"","'","/","@","<"],"resolveProvider":true},"codeActionProvider":{"codeActionKinds":["source.fixAll.ts","source.removeUnused.ts","source.addMissingImports.ts","source.organizeImports.ts","quickfix","refactor"]},"definitionProvider":true,"documentFormattingProvider":true,"documentRangeFormattingProvider":true,"documentHighlightProvider":true,"documentSymbolProvider":true,"executeCommandProvider":{"commands":["_typescript.applyWorkspaceEdit","_typescript.applyCodeAction","_typescript.applyRefactoring","_typescript.configurePlugin","_typescript.organizeImports","_typescript.applyRenameFile","_typescript.goToSourceDefinition"]},"hoverProvider":true,"inlayHintProvider":true,"renameProvider":true,"referencesProvider":true,"signatureHelpProvider":{"triggerCharacters":["(",",","<"],"retriggerCharacters":[")"]},"workspaceSymbolProvider":true,"implementationProvider":true,"typeDefinitionProvider":true,"foldingRangeProvider":true,"semanticTokensProvider":{"documentSelector":null,"legend":{"tokenTypes":["class","enum","interface","namespace","typeParameter","type","parameter","variable","enumMember","property","function","member"],"tokenModifiers":["declaration","static","async","readonly","defaultLibrary","local"]},"full":true,"range":true},"callsProvider":true}}}
2022-11-03T13:02:24.850 helix_lsp::transport [INFO] <- {"capabilities":{"callsProvider":true,"codeActionProvider":{"codeActionKinds":["source.fixAll.ts","source.removeUnused.ts","source.addMissingImports.ts","source.organizeImports.ts","quickfix","refactor"]},"completionProvider":{"resolveProvider":true,"triggerCharacters":[".","\"","'","/","@","<"]},"definitionProvider":true,"documentFormattingProvider":true,"documentHighlightProvider":true,"documentRangeFormattingProvider":true,"documentSymbolProvider":true,"executeCommandProvider":{"commands":["_typescript.applyWorkspaceEdit","_typescript.applyCodeAction","_typescript.applyRefactoring","_typescript.configurePlugin","_typescript.organizeImports","_typescript.applyRenameFile","_typescript.goToSourceDefinition"]},"foldingRangeProvider":true,"hoverProvider":true,"implementationProvider":true,"inlayHintProvider":true,"referencesProvider":true,"renameProvider":true,"semanticTokensProvider":{"documentSelector":null,"full":true,"legend":{"tokenModifiers":["declaration","static","async","readonly","defaultLibrary","local"],"tokenTypes":["class","enum","interface","namespace","typeParameter","type","parameter","variable","enumMember","property","function","member"]},"range":true},"signatureHelpProvider":{"retriggerCharacters":[")"],"triggerCharacters":["(",",","<"]},"textDocumentSync":2,"typeDefinitionProvider":true,"workspaceSymbolProvider":true}}
2022-11-03T13:02:24.850 helix_lsp::transport [INFO] -> {"jsonrpc":"2.0","method":"initialized","params":{}}
2022-11-03T13:02:24.850 helix_view::document [DEBUG] id 1 modified - last saved: 0, current: 0
2022-11-03T13:02:24.850 helix_term::application [DEBUG] received editor event: LanguageServerMessage((0, Notification(Notification { jsonrpc: None, method: "initialized", params: None })))
2022-11-03T13:02:24.850 helix_lsp::transport [INFO] -> {"jsonrpc":"2.0","method":"textDocument/didOpen","params":{"textDocument":{"languageId":"typescript","text":"if(true) &&true\n","uri":"file:///home/itay123/p.ts","version":0}}}
2022-11-03T13:02:24.943 helix_term::application [DEBUG] received editor event: IdleTimer
2022-11-03T13:02:25.543 helix_lsp::transport [INFO] <- {"jsonrpc":"2.0","method":"textDocument/publishDiagnostics","params":{"uri":"file:///home/itay123/p.ts","diagnostics":[{"range":{"start":{"line":0,"character":9},"end":{"line":0,"character":11}},"message":"Expression expected.","severity":1,"code":1109,"source":"typescript"}]}}
2022-11-03T13:02:25.543 helix_term::application [DEBUG] received editor event: LanguageServerMessage((0, Notification(Notification { jsonrpc: Some(V2), method: "textDocument/publishDiagnostics", params: Map({"diagnostics": Array [Object {"code": Number(1109), "message": String("Expression expected."), "range": Object {"end": Object {"character": Number(11), "line": Number(0)}, "start": Object {"character": Number(9), "line": Number(0)}}, "severity": Number(1), "source": String("typescript")}], "uri": String("file:///home/itay123/p.ts")}) })))
2022-11-03T13:02:25.543 helix_view::document [DEBUG] id 1 modified - last saved: 0, current: 0
2022-11-03T13:02:25.801 helix_view::document [DEBUG] id 1 modified - last saved: 0, current: 0
2022-11-03T13:02:25.957 helix_view::document [DEBUG] id 1 modified - last saved: 0, current: 0
2022-11-03T13:02:26.147 helix_view::document [DEBUG] id 1 modified - last saved: 0, current: 0
2022-11-03T13:02:26.292 helix_view::document [DEBUG] id 1 modified - last saved: 0, current: 0
2022-11-03T13:02:26.433 helix_view::document [DEBUG] id 1 modified - last saved: 0, current: 0
2022-11-03T13:02:26.578 helix_view::document [DEBUG] id 1 modified - last saved: 0, current: 0
2022-11-03T13:02:26.738 helix_view::document [DEBUG] id 1 modified - last saved: 0, current: 0
2022-11-03T13:02:26.877 helix_view::document [DEBUG] id 1 modified - last saved: 0, current: 0
2022-11-03T13:02:27.084 helix_term::commands [TRACE] entering insert mode with sel: Selection { ranges: [Range { anchor: 8, head: 9, horiz: None }], primary_index: 0 }, text: "if(true) &&true\n"
2022-11-03T13:02:27.084 helix_view::document [DEBUG] id 1 modified - last saved: 0, current: 0
2022-11-03T13:02:27.085 helix_lsp::transport [INFO] -> {"jsonrpc":"2.0","method":"textDocument/signatureHelp","params":{"position":{"character":8,"line":0},"textDocument":{"uri":"file:///home/itay123/p.ts"}},"id":1}
2022-11-03T13:02:27.093 helix_lsp::transport [INFO] <- {"jsonrpc":"2.0","id":1,"result":null}
2022-11-03T13:02:27.094 helix_lsp::transport [INFO] <- null
2022-11-03T13:02:27.094 helix_view::document [DEBUG] id 1 modified - last saved: 0, current: 0
2022-11-03T13:02:27.484 helix_term::application [DEBUG] received editor event: IdleTimer
2022-11-03T13:02:27.485 helix_view::document [DEBUG] id 1 modified - last saved: 0, current: 0
2022-11-03T13:02:27.920 mio::poll [TRACE] deregistering event source from poller
2022-11-03T13:02:27.921 mio::poll [TRACE] deregistering event source from poller
2022-11-03T13:02:27.921 mio::poll [TRACE] deregistering event source from poller
2022-11-03T13:02:27.921 mio::poll [TRACE] deregistering event source from poller
2022-11-03T13:02:27.921 mio::poll [TRACE] deregistering event source from poller
2022-11-03T13:02:27.922 mio::poll [TRACE] deregistering event source from poller
```

### Platform

Linux

### Terminal Emulator

Kitty

### Helix Version

helix 22.08.1 (8ff92c74)

--- a/helix-core/src/indent.rs
+++ b/helix-core/src/indent.rs
@@ -461,60 +461,58 @@ fn query_indents(
 /// so that the indent computation starts with the correct syntax node.
 fn extend_nodes<'a>(
     node: &mut Node<'a>,
-    deepest_preceding: Option<Node<'a>>,
+    mut deepest_preceding: Node<'a>,
     extend_captures: &HashMap<usize, Vec<ExtendCapture>>,
     text: RopeSlice,
     line: usize,
     tab_width: usize,
 ) {
-    if let Some(mut deepest_preceding) = deepest_preceding {
-        let mut stop_extend = false;
-        while deepest_preceding != *node {
-            let mut extend_node = false;
-            // This will be set to true if this node is captured, regardless of whether
-            // it actually will be extended (e.g. because the cursor isn't indented
-            // more than the node).
-            let mut node_captured = false;
-            if let Some(captures) = extend_captures.get(&deepest_preceding.id()) {
-                for capture in captures {
-                    match capture {
-                        ExtendCapture::PreventOnce => {
-                            stop_extend = true;
-                        }
-                        ExtendCapture::Extend => {
-                            node_captured = true;
-                            // We extend the node if
-                            // - the cursor is on the same line as the end of the node OR
-                            // - the line that the cursor is on is more indented than the
-                            //   first line of the node
-                            if deepest_preceding.end_position().row == line {
+    let mut stop_extend = false;
+
+    while deepest_preceding != *node {
+        let mut extend_node = false;
+        // This will be set to true if this node is captured, regardless of whether
+        // it actually will be extended (e.g. because the cursor isn't indented
+        // more than the node).
+        let mut node_captured = false;
+        if let Some(captures) = extend_captures.get(&deepest_preceding.id()) {
+            for capture in captures {
+                match capture {
+                    ExtendCapture::PreventOnce => {
+                        stop_extend = true;
+                    }
+                    ExtendCapture::Extend => {
+                        node_captured = true;
+                        // We extend the node if
+                        // - the cursor is on the same line as the end of the node OR
+                        // - the line that the cursor is on is more indented than the
+                        //   first line of the node
+                        if deepest_preceding.end_position().row == line {
+                            extend_node = true;
+                        } else {
+                            let cursor_indent = indent_level_for_line(text.line(line), tab_width);
+                            let node_indent = indent_level_for_line(
+                                text.line(deepest_preceding.start_position().row),
+                                tab_width,
+                            );
+                            if cursor_indent > node_indent {
                                 extend_node = true;
-                            } else {
-                                let cursor_indent =
-                                    indent_level_for_line(text.line(line), tab_width);
-                                let node_indent = indent_level_for_line(
-                                    text.line(deepest_preceding.start_position().row),
-                                    tab_width,
-                                );
-                                if cursor_indent > node_indent {
-                                    extend_node = true;
-                                }
                             }
                         }
                     }
                 }
             }
-            // If we encountered some `StopExtend` capture before, we don't
-            // extend the node even if we otherwise would
-            if node_captured && stop_extend {
-                stop_extend = false;
-            } else if extend_node && !stop_extend {
-                *node = deepest_preceding;
-                break;
-            }
-            // This parent always exists since node is an ancestor of deepest_preceding
-            deepest_preceding = deepest_preceding.parent().unwrap();
         }
+        // If we encountered some `StopExtend` capture before, we don't
+        // extend the node even if we otherwise would
+        if node_captured && stop_extend {
+            stop_extend = false;
+        } else if extend_node && !stop_extend {
+            *node = deepest_preceding;
+            break;
+        }
+        // This parent always exists since node is an ancestor of deepest_preceding
+        deepest_preceding = deepest_preceding.parent().unwrap();
     }
 }
 
@@ -612,14 +610,16 @@ pub fn treesitter_indent_for_pos(
     let extend_captures = query_result.extend_captures;
 
     // Check for extend captures, potentially changing the node that the indent calculation starts with
-    extend_nodes(
-        &mut node,
-        deepest_preceding,
-        &extend_captures,
-        text,
-        line,
-        tab_width,
-    );
+    if let Some(deepest_preceding) = deepest_preceding {
+        extend_nodes(
+            &mut node,
+            deepest_preceding,
+            &extend_captures,
+            text,
+            line,
+            tab_width,
+        );
+    }
     let mut first_in_line = get_first_in_line(node, new_line.then(|| byte_pos));
 
     let mut result = Indentation::default();

--- a/helix-core/src/indent.rs
+++ b/helix-core/src/indent.rs
@@ -511,8 +511,12 @@ fn extend_nodes<'a>(
             *node = deepest_preceding;
             break;
         }
-        // This parent always exists since node is an ancestor of deepest_preceding
-        deepest_preceding = deepest_preceding.parent().unwrap();
+        // If the tree contains a syntax error, `deepest_preceding` may not
+        // have a parent despite being a descendant of `node`.
+        deepest_preceding = match deepest_preceding.parent() {
+            Some(parent) => parent,
+            None => return,
+        }
     }
 }
 

